docs/core/dependency-loading/understanding-assemblyloadcontext.md

```diff
@@ -0,0 +1,113 @@
+
+## How are dynamic dependencies isolated?
+
+Each <xref:System.Runtime.Loader.AssemblyLoadContext> instance represents a unique scope for <xref:System.Reflection.Assembly> instances and <xref:System.Type> definitions.
+
+There's no binary isolation between these dependencies. They're only isolated by not finding each other by name.
+
+In each <xref:System.Runtime.Loader.AssemblyLoadContext>:
+    - <xref:System.Reflection.AssemblyName.Name?displayProperty=nameWithType> may refer to a different <xref:System.Reflection.Assembly> instance.
+    - <xref:System.Type.GetType%2A?displayProperty=nameWithType> may return a different type instance for the same type `name`.
+
+## How are dependencies shared?
+
+Dependencies can easily be shared between <xref:System.Runtime.Loader.AssemblyLoadContext> instances. The general model is for one <xref:System.Runtime.Loader.AssemblyLoadContext> to load a dependency.  The other shares the dependency by using a reference to the loaded assembly.
+
+This sharing is required of the runtime assemblies. These assemblies can only be loaded into the <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType>. The same is required for frameworks like `ASP.NET`, `WPF`, or `WinForms`.
+
+It's recommended that shared dependencies should be loaded into <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType>. This sharing is the common design pattern.
+
+Sharing is implemented in the coding of the custom <xref:System.Runtime.Loader.AssemblyLoadContext> instance. <xref:System.Runtime.Loader.AssemblyLoadContext> has various events and virtual functions that can be overridden. When any of these functions return a reference to an <xref:System.Reflection.Assembly> instance that was loaded in another <xref:System.Runtime.Loader.AssemblyLoadContext> instance, the <xref:System.Reflection.Assembly> instance is shared. The standard load algorithm defers to <xref:System.Runtime.Loader.AssemblyLoadContext.Default?displayProperty=nameWithType> for loading to simplify the common sharing pattern.  See [Managed assembly loading algorithm](loading-managed.md).
+
+## Complications
+
+### Type conversion issues
+
+When two <xref:System.Runtime.Loader.AssemblyLoadContext> instances contain type definitions with the same `name`, they're not the same type. They're the same type if and only if they come from the same <xref:System.Reflection.Assembly> instance.
+
+To complicate matters, exception messages about these mismatched types can be confusing. The types are referred to in the exception messages by their simple type names. The common exception message in this case would be of the form:
```

Comment:
This is a bug to fix. We have added a name to AssemblyLoadContext to be able to produce good diagnostics for cases like these.

Reply:
I am not sure we have time to fix for 3.0 RTM.  
I had thought we weren't allowed to fix this because changing the message was a breaking change.

Reply:
3.0 is already way too late... unfortunately. We can try to get this to 3.1, but I doubt it would make it (it's a nice to have).
We should do this for 5.0 - which will completely erase any concerns about braking changes. I personally don't think changing exception messages is treated as a breaking change (CoreFX does this all the time) - we just must not change the exception type (that would be considered a breaking change).

Reply:
I agree that fixing this won’t meet the bar for 3.0. It is fine to fix this for .NET 5. Improving exception messages is acceptable change.